File: apps/portal/web/src/api/stormservice.ts
import { apiFetch, buildQuery, type ListParams, type Pagination } from "./client";

export interface StormServiceListItem {
  name: string;
  namespace: string;
  replicas: number;
  readyReplicas: number;
  rolesCount: number;
  stateful: boolean;
  createdAt: string;
}

export interface StormServiceListResponse {
  items: StormServiceListItem[];
  pagination: Pagination;
}

export interface StormRoleSpec {
  name: string;
  replicas: number;
  image: string;
  cpu?: string;
  memory?: string;
  gpu?: number;
}

export interface StormServiceCreateRequest {
  name: string;
  namespace: string;
  replicas?: number;
  stateful?: boolean;
  roles: StormRoleSpec[];
}

export interface StormServiceDetail {
  name: string;
  namespace: string;
  stateful: boolean;
  status: { replicas: number; readyReplicas: number };
  createdAt: string;
}

export const stormServiceApi = {
  list: (params?: ListParams) =>
    apiFetch<StormServiceListResponse>(`/stormservices${buildQuery(params)}`),
  get: (ns: string, name: string) =>
    apiFetch<StormServiceDetail>(`/stormservices/${ns}/${name}`),
  create: (data: StormServiceCreateRequest) =>
    apiFetch<StormServiceDetail>("/stormservices", { method: "POST", body: JSON.stringify(data) }),
  delete: (ns: string, name: string) =>
    apiFetch<void>(`/stormservices/${ns}/${name}`, { method: "DELETE" }),
};

File: apps/portal/api/types/stormservice.go
package types

import "time"

type StormServiceCreateRequest struct {
	Name      string          `json:"name" binding:"required"`
	Namespace string          `json:"namespace" binding:"required"`
	Replicas  *int32          `json:"replicas,omitempty"`
	Stateful  bool            `json:"stateful,omitempty"`
	Roles     []StormRoleSpec `json:"roles" binding:"required"`
}

type StormServiceUpdateRequest struct {
	Replicas *int32          `json:"replicas,omitempty"`
	Roles    []StormRoleSpec `json:"roles,omitempty"`
}

type StormRoleSpec struct {
	Name     string `json:"name" binding:"required"`
	Replicas int32  `json:"replicas"`
	Image    string `json:"image" binding:"required"`
	CPU      string `json:"cpu,omitempty"`
	Memory   string `json:"memory,omitempty"`
	GPU      int32  `json:"gpu,omitempty"`
}

type StormServiceListItem struct {
	Name          string    `json:"name"`
	Namespace     string    `json:"namespace"`
	Replicas      int32     `json:"replicas"`
	ReadyReplicas int32     `json:"readyReplicas"`
	RolesCount    int       `json:"rolesCount"`
	Stateful      bool      `json:"stateful"`
	CreatedAt     time.Time `json:"createdAt"`
}

type StormServiceListResponse struct {
	Items      []StormServiceListItem `json:"items"`
	Pagination Pagination             `json:"pagination"`
}

type StormServiceDetailResponse struct {
	Name       string                   `json:"name"`
	Namespace  string                   `json:"namespace"`
	Stateful   bool                     `json:"stateful"`
	RolesCount int                      `json:"rolesCount"`
	Status     StormServiceDetailStatus `json:"status"`
	CreatedAt  time.Time                `json:"createdAt"`
}

type StormServiceDetailStatus struct {
	Replicas      int32 `json:"replicas"`
	ReadyReplicas int32 `json:"readyReplicas"`
}
